pkg/cmd/cli/cmd/volume.go

```diff
@@ -0,0 +1,602 @@
+	if len(v.Name) == 0 {
+		for _, c := range containers {
+			c.VolumeMounts = []kapi.VolumeMount{}
+		}
+		spec.Volumes = []kapi.Volume{}
+	} else {
+		for _, c := range containers {
+			for i, m := range c.VolumeMounts {
+				if v.Name == m.Name {
+					c.VolumeMounts = append(c.VolumeMounts[:i], c.VolumeMounts[i+1:]...)
+					break
+				}
+			}
+		}
+
+		// Remove volume if no container is using it
+		found := false
+		for _, c := range skippedContainers {
+			for _, m := range c.VolumeMounts {
+				if v.Name == m.Name {
+					found = true
+					break
+				}
+			}
+			if found {
+				break
+			}
+		}
+		if !found {
+			for i, vol := range spec.Volumes {
+				if v.Name == vol.Name {
+					spec.Volumes = append(spec.Volumes[:i], spec.Volumes[i+1:]...)
+					break
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (v *VolumeOptions) listVolumeForSpec(spec *kapi.PodSpec, info *resource.Info) error {
+	fmt.Fprintf(v.Writer, "# %s %s, volumes:\n", info.Mapping.Resource, info.Name)
+	checkName := (len(v.Name) > 0)
+	for _, vol := range spec.Volumes {
+		if checkName && v.Name != vol.Name {
+			continue
+		}
+		fmt.Fprintf(v.Writer, "%s\n", vol.Name)
+	}
+
+	containers, _ := v.selectContainers(spec.Containers, v.Containers)
+	for _, c := range containers {
+		fmt.Fprintf(v.Writer, "\t# container %s, volume mounts:\n", c.Name)
+		for _, m := range c.VolumeMounts {
+			if checkName && v.Name != m.Name {
+				continue
+			}
+			fmt.Fprintf(v.Writer, "\t%s %s\n", m.Name, m.MountPath)
+		}
+		fmt.Fprintf(v.Writer, "\n")
+	}
+	fmt.Fprintf(v.Writer, "\n")
+
+	return nil
+}
+
+func (v *VolumeOptions) selectContainers(containers []kapi.Container, spec string) ([]*kapi.Container, []*kapi.Container) {
+	out := []*kapi.Container{}
+	skipped := []*kapi.Container{}
+	for i, c := range containers {
+		if v.selectString(c.Name, spec) {
+			out = append(out, &containers[i])
+		} else {
+			skipped = append(skipped, &containers[i])
+		}
+	}
+	return out, skipped
+}
+
+// selectString returns true if the provided string matches spec, where spec is a string with
+// a non-greedy '*' wildcard operator.
+// TODO: turn into a regex and handle greedy matches and backtracking.
+func (v *VolumeOptions) selectString(s, spec string) bool {
+	if spec == "*" {
+		return true
+	}
+	if !strings.Contains(spec, "*") {
+		return s == spec
+	}
+
+	pos := 0
+	match := true
+	parts := strings.Split(spec, "*")
+	for i, part := range parts {
+		if len(part) == 0 {
+			continue
+		}
+		next := strings.Index(s[pos:], part)
+		switch {
+		// next part not in string
+		case next < pos:
+			fallthrough
+		// first part does not match start of string
+		case i == 0 && pos != 0:
+			fallthrough
+		// last part does not exactly match remaining part of string
+		case i == (len(parts)-1) && len(s) != (len(part)+next):
+			match = false
+			break
+		default:
+			pos = next
+		}
+	}
+	return match
+}
+
+func (v *VolumeOptions) handleUpdateError(err error) {
```

Comment:
This method shouldn't be part of volumeOptions because it doesn't need any local fields.